### DISCUSSION TITLE
Very high latency in MVT layer when loading tiles.

### DISCUSSION THREAD

**Discussion by smcallis** [visgl/deck.gl#9089] [Q&A]
I benchmarked my deck.gl in Chrome and I noticed some latency oddness I'm hoping someone can help me with.  I'm using an MVTLayer and the tiles in this particular case aren't that big, about 800KB, but when I load the page it can take up to 20s for the tiles to become visible, long after they've actually finished downloading:

![image](https://github.com/user-attachments/assets/a93eb043-4b49-45cb-b4b4-7aaa895ea610)

It looks like deck.gl doesn't do anything until a random AnimationFrameFired occurs, then it initializes the layers:

![image](https://github.com/user-attachments/assets/969148e5-0ba8-488a-9596-efa5b6576b87)

Is there any way to force this to happen as sooner as the tile data is ready?

**Comment by smcallis** [comment]
Perhaps the issue is in my custom fetch function if that helps:

```

         fetch: async (url, { layer, loaders, loadOptions, signal }) => {
           // MVTLayer stores a special loader on its properties.
           loaders = loaders || layer.props.loaders

           const retrieve = async function(options={}) {
             return fetch(url, {
               credentials: "include",
               ...options
             });
           }

           const parseData = async function(data) {
             return window.loaders.parse(data, loaders, loadOptions);
           }

           // Try to retrieve the data.  If we get an error, retry without CORS
           // enabled to load any authentication cookies that are needed, and
           // retry.
           return new Promise((resolve, reject) => {
             retrieve()
               .then(data => resolve(parseData(data)))
               .catch(error => {
                 console.log(`Fetch failed ${error}, retrying`);
                 retrieve({mode: 'no-cors'})
                   .then(data => {
                     retrieve()
                       .then(data => resolve(parseData(data)))
                       .catch(reject); })
                   .catch(reject);
                 });
           });
         },
```

**Comment by smcallis** [comment]
Actually I think it's the data parser: 

![image](https://github.com/user-attachments/assets/f746d8c6-1efc-4f52-997f-ad2da8e9e802)

**Comment by ibgreen** [comment]
I believe the MVTLayer supports a binary mode flag, which would avoid the conversion to and from GeoJSON.

**Reply by smcallis** [reply]
My understanding is that `binary: true` is the default, but I just tried adding it and it didn't make a difference.

**Reply by ibgreen** [reply]
Looking at your call stack it looks like the MVTLoader is converting binary to GeoJSON. You could check if that is still happening after setting the prop. If so we could eliminate that as the issue.

**Reply by smcallis** [reply]
Done, this is the current stack:

![image](https://github.com/user-attachments/assets/782c2586-5289-4848-bcb3-d2b5cdb4a0fc)

I can't even find the classifyRings2 function...

**Reply by smcallis** [reply]
Nevermind it's just that there's a separate classifyRings function so it got renamed when bundling.  It's a BinaryVectorTileFeature so I think it's reading it as binary, but the ring classification is just a killer for polygons.  When I load a tile full of points it's not a problem.

**Reply by smcallis** [reply]
For perspective, the tile is 11608 polygons with a total of 1,855,149 points, about 160 points/polygon.  So not small but not 15 seconds worth of processing either.

**Comment by smcallis** [comment]
I rewrote the getPolygonSignedArea() function without the loop cleverness to help the JIT and it's night and day faster, this is the code:

```
 function getPolygonSignedArea(points, options = {}) {
    const {
      end = points.length,
      plane = "xy"
    } = options;

    const dim = options.size || 2;
    const i0 = DimIndex[plane[0]];
    const i1 = DimIndex[plane[1]];

    let Area = function(curr, prev) {
      return
        (points[curr + i0] - points[prev + i0])*
        (points[curr + i1] - points[prev + i1]);
    };

    let area2 = Area(0, end-dim);
    for (let i = dim; i < end; i += dim) {
      area2 += Area(i, i - dim);
    }
    return area2/2;
```

And this is the stack trace now, classifyRings is in the millisecond range now and fillArrays is the bottleneck:

![image](https://github.com/user-attachments/assets/4ef7fa64-bb00-4c8b-b1ac-43eab25459a8)

**Reply by ibgreen** [reply]
@smcallis This looks very promising indeed. I made a change to math.gl to use your math however I can't replicate the results in our benchmarks, if you look at the `Bench` section in:
https://github.com/visgl/math.gl/actions/runs/10433413865/job/28895112777?pr=19
admittedly those bench tests run under Node 20, not in the browser/puppeteer so it may not be apples to apples.

**Reply by smcallis** [reply]
I'm definitely far from an expert in appeasing the JIT, all I can really say is it got faster for _me_, mileage may vary for others.   It looked like the getPolygonSignedArea() code in HEAD in math.gl was a little different than the one that was bundled with the latest deck.gl, so that might affect things too, maybe it's been fixed already?
